### PR TITLE
Documentation for EKS support using private VPCs

### PR DESCRIPTION
**Pull request by buzzsurfr**
*Issue #, if available:*

*Description of changes:*
Add/update documentation discussing the options for creating, deploying, and operating an EKS cluster in a private VPC without outbound internet access.

Contributed by @buzzsurfr, @rafaelpereyra, and Jack Iu

Related to the following issues:
* aws/containers-roadmap#586
* aws/containers-roadmap#666
* aws/containers-roadmap#679

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: doc_source/create-cluster.md [hunk 1]
```diff
@@ -89,6 +89,8 @@ If you select subnets that were created before 03/26/2020 using one of the Amazo
 **Important**  
 The worker node AWS CloudFormation template modifies the security group that you specify here, so **Amazon EKS strongly recommends that you use a dedicated security group for each cluster control plane \(one per cluster\)**\. If this security group is shared with other resources, you might block or disrupt connections to those resources\.
    + **Endpoint private access** – Choose whether to enable or disable private access for your cluster's Kubernetes API server endpoint\. If you enable private access, Kubernetes API requests that originate from within your cluster's VPC use the private VPC endpoint\. For more information, see [Amazon EKS cluster endpoint access control](cluster-endpoint.md)\.
+**Important**  
+If you created a VPC with only private subnets, you must enable **Endpoint private access**.
```
Comment (prithviramesh) [inline doc_source/create-cluster.md:93]: I don't think this is true

Comment (prithviramesh) [reply doc_source/create-cluster.md:93]: A VPC with private subnets should still be allowed to have Endpoint Public Access set IIUC, it's if the VPC has *no internet access*, then the cluster would need to have Endpoint private access enabled.

Comment (prithviramesh) [reply doc_source/create-cluster.md:93]: Unless the "only" in "VPC with only private subnets" means there are no NATs that the "private" subnets are associated with.

Comment (prithviramesh) [reply doc_source/create-cluster.md:93]: the IGW would technically have to reside in a public subnet, so I think this statement may actually be correct.

File: doc_source/create-public-private-vpc.md [hunk 2]
```diff
@@ -16,6 +16,11 @@ Before 03/26/2020 – Public IPv4 addresses are not automatically assigned by pu
...
+
+**Important**
+There are additional requirements if the VPC does not have outbound internet access, such as via a NAT Instance, NAT Gateway, VPN, or Direct Connect\. You must bypass the EKS cluster introspection by providing the cluster certificate authority and cluster API endpoint to the worker nodes\. You also may need to configure VPC endpoints listed in [VPC endpoints for private clusters](private-clusters#vpc-endpoints-for-private-clusters)\.
```
Comment (prithviramesh) [inline doc_source/create-public-private-vpc.md:22]: The VPC endpoint consideration is a requirement if they are attempting to run Fargate pods in Subnets (and thus a VPC) that doesn't have outbound internet access. Can we call that out here as well?

File: doc_source/private-clusters.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+This chapter discusses running Amazon EKS in a private cluster without outbound internet access, commonly referred to as "air-gapped".
+
+## Private cluster considerations<a name="private-cluster-considerations"></a>
+
+Here's some things to consider when running Amazon EKS in a private cluster without outbound internet access\.
++ Any container image that is not in Amazon ECR or a registry inside the VPC will need to be copied to Amazon ECR or a registry inside the VPC in order to be pulled\. See [Creating local copies of container images](#creating-local-copies-of-container-images)\.
++ **Endpoint private access** is required in order for the worker nodes to register with the cluster endpoint\. **Endpoint public access** is optional\.
++ You must bypass the EKS cluster introspection by providing the cluster certificate authority and cluster API endpoint to the worker nodes\.
++ You must include the VPC endpoints found at [VPC endpoints for private clusters](#vpc-endpoints-for-private-clusters)\.
++ Managed node groups is not supported with private clusters\.
```
Comment (abeer91) [inline doc_source/private-clusters.md:12]: This is supported with the proposed template

File: doc_source/private-clusters.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+## VPC endpoints for private clusters<a name="vpc-endpoints-for-private-clusters"></a>
+
+The following [VPC endpoints](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-endpoints.html) may be required.
+
++ **com.amazonaws.REGION.ec2**
++ **com.amazonaws.REGION.ecr.api**
++ **com.amazonaws.REGION.ecr.dkr**
++ **com.amazonaws.REGION.s3** for pulling container images
++ **com.amazonaws.REGION.logs**
```
Comment (wangamel) [inline doc_source/private-clusters.md:76]: Basically, for launching fargate pods, endpoints for ecr, s3 and sts are required to have. All other endpoints are optional and depend on your needs.

File: doc_source/private-clusters.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+This chapter discusses running Amazon EKS in a private cluster without outbound internet access, commonly referred to as "air-gapped".
+
+## Private cluster considerations<a name="private-cluster-considerations"></a>
+
+Here's some things to consider when running Amazon EKS in a private cluster without outbound internet access\.
++ Any container image that is not in Amazon ECR or a registry inside the VPC will need to be copied to Amazon ECR or a registry inside the VPC in order to be pulled\. See [Creating local copies of container images](#creating-local-copies-of-container-images)\.
++ **Endpoint private access** is required in order for the worker nodes to register with the cluster endpoint\. **Endpoint public access** is optional\.
++ You must bypass the EKS cluster introspection by providing the cluster certificate authority and cluster API endpoint to the worker nodes\.
++ You must include the VPC endpoints found at [VPC endpoints for private clusters](#vpc-endpoints-for-private-clusters)\.
```
Comment (wangamel) [inline doc_source/private-clusters.md:11]: To launch fargate pods on offline private subnets, set up endpoints for ecr, s3 and sts.

File: doc_source/private-clusters.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+   --apiserver-endpoint {CLUSTER_ENDPOINT} --b64-cluster-ca {CLUSTER_CERTIFICATE_AUTHORITY}
+   ```
+
++ [IAM roles for service accounts](iam-roles-for-service-accounts) is supported\. You must include the STS VPC endpoint\.
++ The [Amazon EBS CSI driver](ebs-csi) is supported\. When deploying, the [kustomization.yaml](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/deploy/kubernetes/overlays/stable/kustomization.yaml) file must be changed to set the container images to use the same region as the EKS cluster\.
++ The [Amazon EFS CSI driver](efs-csi) is supported\. When deploying, the [kustomization.yaml](https://github.com/kubernetes-sigs/aws-efs-csi-driver/blob/master/deploy/kubernetes/overlays/stable/kustomization.yaml) file must be changed to set the container images to use the same region as the EKS cluster\.
++ The [Amazon FSx for Lustre CSI driver](fsx-csi) is not supported\.
++ The `aws-auth` ConfigMap must be created from within the VPC\.
++ The [ALB Ingress Controller on Amazon EKS](alb-ingress) does not work in private clusters\.
++ [AWS Fargate](fargate) is supported with private clusters. You must include the STS VPC endpoint\.
```
Comment (prithviramesh) [inline doc_source/private-clusters.md:30]: Like Amelia was pointing out, STS/S3/ECR VPC endpoints are all required for EKS Fargate to function properly in an "air-gapped" environment

**Comment by jimdial-aws** [comment]
Closing due to merge conflicts now, but this content is published.

**Pull request closed without merging**